lib/Doctrine/Migrations/Configuration/Loader/XmlFileLoader.php

```diff
@@ -0,0 +1,126 @@
+use SimpleXMLElement;
+use const DIRECTORY_SEPARATOR;
+use const LIBXML_NOCDATA;
+use function assert;
+use function file_exists;
+use function file_get_contents;
+use function libxml_clear_errors;
+use function libxml_use_internal_errors;
+use function simplexml_load_string;
+use function strtr;
+
+/**
+ * @internal
+ */
+final class XmlFileLoader extends AbstractFileLoader
+{
+    /** @var ArrayLoader */
+    private $arrayLoader;
+
+    public function __construct(?ArrayLoader $arrayLoader = null)
+    {
+        $this->arrayLoader = $arrayLoader ?: new ArrayLoader();
+    }
+
+    /**
+     * @param mixed|string $file
+     */
+    public function load($file) : Configuration
+    {
+        if (! file_exists($file)) {
+            throw FileNotFound::new();
+        }
+
+        $this->validateXML($file);
+
+        $rawXML = file_get_contents($file);
+        assert($rawXML !== false);
+
+        $root = simplexml_load_string($rawXML, SimpleXMLElement::class, LIBXML_NOCDATA);
+        assert($root !== false);
+
+        $config = $this->extractParameters($root);
+
+        if (isset($config['all_or_nothing'])) {
+            $config['all_or_nothing'] = BooleanStringFormatter::toBoolean(
+                $config['all_or_nothing'],
+                false
+            );
+        }
+        if (isset($config['migrations_paths'])) {
+            $config['migrations_paths'] = $this->getDirectoryRelativeToFile(
+                $file,
+                $config['migrations_paths']
+            );
+        }
+
+        return $this->arrayLoader->load($config);
+    }
+
+    /**
+     * @return mixed[]
+     */
+    private function extractParameters(SimpleXMLElement $root, bool $loopOverNodes = true) : array
+    {
+        $config = [];
+
+        $itemsToCheck = $loopOverNodes ? $root->children() : $root->attributes();
+
+        if (! ($itemsToCheck instanceof SimpleXMLElement)) {
+            return $config;
+        }
+        foreach ($itemsToCheck as $node) {
+            $nodeName = strtr($node->getName(), '-', '_');
+            if ($nodeName === 'migrations_paths') {
+                $config['migrations_paths'] = [];
+                foreach ($node->{'path'} as $pathNode) {
+                    $config['migrations_paths'][(string) $pathNode['namespace']] = (string) $pathNode;
+                }
+            } elseif ($nodeName === 'storage' && $node->{'table-storage'} instanceof SimpleXMLElement) {
+                $config['table_storage'] = $this->extractParameters($node->{'table-storage'}, false);
+            } elseif ($nodeName === 'migrations') {
+                $config['migrations'] = [];
+                foreach ($node->{'migration'} as $pathNode) {
+                    $config['migrations'][] = (string) $pathNode;
+                }
+            } else {
+                $config[$nodeName] = (string) $node;
+            }
+        }
+
+        return $config;
+    }
+
+    private function validateXML(string $file) : void
```

Comment:
For consistency, should this be named

```suggestion
    private function validateXml(string $file) : void
```

?